### PR TITLE
[mlir][AMDGPU] Add support for AMD f16 math library calls

### PR DESCRIPTION
**Pull request by dhernandez0**
In this PR we add support for AMD f16 math library calls (`__ocml_*_f16`)

CC: @krzysz00 @manupak

### CODE REVIEW

File: mlir/lib/Conversion/MathToROCDL/MathToROCDL.cpp [hunk 1]
```diff
@@ -63,59 +63,61 @@ void mlir::populateMathToROCDLConversionPatterns(LLVMTypeConverter &converter,
...
+                                  "__ocml_cos_f64", "__ocml_cos_f16");
   populateOpPatterns<math::CoshOp>(converter, patterns, "__ocml_cosh_f32",
-                                   "__ocml_cosh_f64");
+                                   "__ocml_cosh_f64", "__ocml_cosh_f16");
   populateOpPatterns<math::SinhOp>(converter, patterns, "__ocml_sinh_f32",
-                                   "__ocml_sinh_f64");
-  populateOpPatterns<math::ExpOp>(converter, patterns, "", "__ocml_exp_f64");
+                                   "__ocml_sinh_f64", "__ocml_sinh_f16");
+  populateOpPatterns<math::ExpOp>(converter, patterns, "", "__ocml_exp_f64",
+                                  "__ocml_exp_f16");
```
Comment (nirvedhmeshram) [inline mlir/lib/Conversion/MathToROCDL/MathToROCDL.cpp:90]: I am not sure if we want to call to library here for f16 if we are not calling to library for f32?

File: mlir/lib/Conversion/MathToROCDL/MathToROCDL.cpp [hunk 2]
```diff
@@ -63,59 +63,61 @@ void mlir::populateMathToROCDLConversionPatterns(LLVMTypeConverter &converter,
...
+                                   "__ocml_exp2_f64", "__ocml_exp2_f16");
   populateOpPatterns<math::ExpM1Op>(converter, patterns, "__ocml_expm1_f32",
-                                    "__ocml_expm1_f64");
+                                    "__ocml_expm1_f64", "__ocml_expm1_f16");
   populateOpPatterns<math::FloorOp>(converter, patterns, "__ocml_floor_f32",
-                                    "__ocml_floor_f64");
-  populateOpPatterns<math::LogOp>(converter, patterns, "", "__ocml_log_f64");
+                                    "__ocml_floor_f64", "__ocml_floor_f16");
+  populateOpPatterns<math::LogOp>(converter, patterns, "", "__ocml_log_f64",
+                                  "__ocml_log_f16");
```
Comment (nirvedhmeshram) [inline mlir/lib/Conversion/MathToROCDL/MathToROCDL.cpp:98]: I am not sure if we want to call to library here for f16 if we are not calling to library for f32?

**Review by nirvedhmeshram — CHANGES_REQUESTED** [review]
Regarding the two f16 calls that you are adding where f32 calls are not present, I am fine with whichever choice makes sense for the hardware / AMDGPU backend but the following needs to be updated accordingly,
https://github.com/llvm/llvm-project/blob/main/mlir/lib/Conversion/GPUToROCDL/LowerGpuOpsToROCDLOps.cpp#L338
It controls if for a given type we will make a llvm op or the library call. Otherwise which one happens may vary based on which pattern gets applied first or in the worst case we end up with the op not getting lowered at all and causing a crash.

File: mlir/lib/Conversion/MathToROCDL/MathToROCDL.cpp [hunk 3]
```diff
@@ -39,16 +39,17 @@ template <typename OpTy>
...
   patterns.add<OpToFuncCallLowering<OpTy>>(converter, f32Func, f64Func,
-                                           f32ApproxFunc);
+                                           f32ApproxFunc, f16Func);
 }
 
 void mlir::populateMathToROCDLConversionPatterns(LLVMTypeConverter &converter,
                                                  RewritePatternSet &patterns) {
   // Handled by mathToLLVM: math::AbsIOp
-  // Handled by mathToLLVM: math::AbsFIOp
+  // Handled by mathToLLVM: math::AbsFOp
```
[the remark this answers is not in the document]
Comment (nirvedhmeshram) [reply mlir/lib/Conversion/MathToROCDL/MathToROCDL.cpp:51]: There is some dicussion on why it is the way it is here
https://github.com/llvm/llvm-project/pull/102971

File: mlir/lib/Conversion/GPUCommon/OpToFuncCallLowering.h [hunk 4]
```diff
@@ -89,7 +91,14 @@ struct OpToFuncCallLowering : public ConvertOpToLLVMPattern<SourceOp> {
...
     Type type = operand.getType();
-    if (!isa<Float16Type>(type))
+    if (!isa<FloatType>(type))
+      return operand;
+
+    // if there's a f16 function, no need to cast f16 values
+    if (!f16Func.empty() && isa<Float16Type>(type))
+      return operand;
+
+    if (isa<Float64Type>(type) || isa<Float32Type>(type))
```
Comment (nirvedhmeshram) [inline mlir/lib/Conversion/GPUCommon/OpToFuncCallLowering.h:101]: wont this also cast for other non f16/f32/f64 types?
I think you just want to modify the exisiting check to 
```
    // if there's a f16 function, no need to cast f16 values
    if (!isa<Float16Type>(type) || !f16Func.empty())
      return operand;
 ```
and leave rest of the logic as is?

File: mlir/lib/Conversion/MathToROCDL/MathToROCDL.cpp (hunk 2 above)
Comment (krzysz00) [reply mlir/lib/Conversion/MathToROCDL/MathToROCDL.cpp:98]: This log implementation is definitely non-trivial
```
; Function Attrs: mustprogress nofree norecurse nosync nounwind willreturn memory(none) define linkonce_odr protected half @__ocml_log_f16(half noundef %0) local_unnamed_addr #0 {
  %2 = fpext half %0 to float
  %3 = tail call float @llvm.amdgcn.log.f32(float %2)
  %4 = fmul float %3, 0x3FE62E4300000000
  %5 = fptrunc float %4 to half
  ret half %5
}; Function Attrs: mustprogress nofree norecurse nosync nounwind willreturn memory(none) define linkonce_odr protected half @__ocml_log_f16(half noundef %0) local_unnamed_addr #0 {
  %2 = fpext half %0 to float
  %3 = tail call float @llvm.amdgcn.log.f32(float %2)
  %4 = fmul float %3, 0x3FE62E4300000000
  %5 = fptrunc float %4 to half
  ret half %5
}
```
so we might want to leave it in (the f32 one just calls an intrinsic)

File: mlir/lib/Conversion/GPUCommon/OpToFuncCallLowering.h (hunk 4 above)
Comment (krzysz00) [reply mlir/lib/Conversion/GPUCommon/OpToFuncCallLowering.h:101]: But we do want to cast, for, say, bf16, I think

Or at least it feels reasonable to do that.

Also, nit, combine the `isa<>`s together

Comment (nirvedhmeshram) [reply mlir/lib/Conversion/GPUCommon/OpToFuncCallLowering.h:101]: That would be different functionality then what the function currently does but yes seems reasonable to do, probably want to add a test for it though. showing the casting happening for bf16 for example.

Comment (nirvedhmeshram) [reply mlir/lib/Conversion/GPUCommon/OpToFuncCallLowering.h:101]: Also this comment will need updating
https://github.com/llvm/llvm-project/blob/main/mlir/lib/Conversion/GPUCommon/OpToFuncCallLowering.h#L23

File: mlir/lib/Conversion/MathToROCDL/MathToROCDL.cpp (hunk 1 above)
Comment (dhernandez0) [reply mlir/lib/Conversion/MathToROCDL/MathToROCDL.cpp:90]: Same thing as @krzysz00 comment regarding `__ocml_log_f16`, it seems `__ocml_exp_f16` is not just calling the intrinsic (while `__ocml_exp_f32` just calls its intrinsic). So, we might want to keep it.

File: mlir/lib/Conversion/GPUCommon/OpToFuncCallLowering.h (hunk 4 above)
Comment (dhernandez0) [reply mlir/lib/Conversion/GPUCommon/OpToFuncCallLowering.h:101]: Hi @nirvedhmeshram, the comment is already updated. Thanks for your suggestion, I'll add some tests.

**Comment by dhernandez0** [comment]
> Regarding the two f16 calls that you are adding where f32 calls are not present, I am fine with whichever choice makes sense for the hardware / AMDGPU backend but the following needs to be updated accordingly, https://github.com/llvm/llvm-project/blob/main/mlir/lib/Conversion/GPUToROCDL/LowerGpuOpsToROCDLOps.cpp#L338 It controls if for a given type we will make a llvm op or the library call. Otherwise which one happens may vary based on which pattern gets applied first or in the worst case we end up with the op not getting lowered at all and causing a crash.

Thanks for your review! I've updated LowerGpuOpsToROCDLOps.cpp as suggested.

**Review by nirvedhmeshram — APPROVED** [review]
LGTM!

**Pull request merged**